### PR TITLE
add example for App::data_factory

### LINKED ISSUE #256: create example usage of async data_factory
- https://github.com/elertan/microservice-rust-test/blob/722336d5acec9e2f33d4e9b4b46cc52baa1b3ee8/logging_service/src/main.rs
- https://github.com/TechEmpower/FrameworkBenchmarks/blob/0d4554867d07cf2a103ddf0154682162e6c3e827/frameworks/Rust/actix/src/main_pg.rs#L128

### PR DESCRIPTION
**Pull request by fakeshadow**
`App::data_factory` can be used to optimize async tasks between works

closes #256

### CODE REVIEW

File: async_data_factory/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+use redis_tang::{Builder, Pool, RedisManager};
+
+// change according to your redis setting.
+const REDIS_URL: &str = "redis://127.0.0.1";
+
+#[actix_rt::main]
+async fn main() -> std::io::Result<()> {
+    let num_cpus = num_cpus::get();
+
+    // a shared redis pool for work load comparision.
```
Comment (JohnTitor) [inline async_data_factory/src/main.rs:13]:
```suggestion
    // a shared redis pool for work load comparison.
```

File: Cargo.toml [hunk 2]
```diff
@@ -4,6 +4,7 @@ members = [
    "async_ex1",
    "async_ex2",
    "async_pg",
+   "async_data_factory",
```
Comment (JohnTitor) [inline Cargo.toml:7]: Should be ordered.

File: async_data_factory/README.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
+## Usage:
+This is an example on constructing async state with `App::data_factory`
+
+## Reason:
+`data_factory` would make sense in these situations:
+- When async state not necessarily have to be shared between workers/threads.
+
+- When async state would spawn tasks on `actix-rt`. If we centralized the state there could be a possibilitythe tasks get a very unbalanced distribution on the workers/threads
```
Comment (JohnTitor) [inline async_data_factory/README.md:8]:
```suggestion
- When async state would spawn tasks on `actix-rt`. If we centralized the state there could be a possibility
the tasks get a very unbalanced distribution on the workers/threads
```

`possibilitythe` -> `possibility the`

File: async_data_factory/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            .data_factory(|| {
+                /*
+                    App::data_factory would accept a future as return type and poll the future when App is initialized.
+
+                    The Output of the future must be Result<T, E> and T will be the transformed to App::Data<T> that can be extracted from handler/request.
+                    (The E will only be used to trigger a log::error.)
+
+                    This data is bound to worker thread and you get an instance of it for every worker of the HttpServer.(hence the name data_factory)
+                    *. It is NOT shared between workers(unless the underlying data is a smart pointer like Arc<T>).
+                */
```
Comment (JohnTitor) [inline async_data_factory/src/main.rs:30]: It'd be great if we could split lines to 120 characters or less.

File: async_data_factory/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    .bind("127.0.0.1:8080")?
+    .workers(num_cpus)
+    .run()
+    .await
+}
+
+// this pool is shared between workers.
+// we have all redis connections spawned tasks on main thread therefore it puts too much pressure on one thread.
+// *. This is the case for redis::aio::MultiplexedConnection and it may not apply to other async redis connection type.
+async fn pool_shared_prebuilt(pool: Data<Wrapper>) -> &'static str {
```
Comment (robjtede) [inline async_data_factory/src/main.rs:51]: use route macros on routes

File: async_data_factory/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        .query_async::<_, ()>(&mut client)
+        .await
+        .unwrap();
+
+    "Done"
+}
+
+// some boiler plate for create redis pool
+#[derive(Clone)]
+struct Wrapper(Pool<RedisManager>);
```
Comment (robjtede) [inline async_data_factory/src/main.rs:79]: `RedisWrapper` for clarity in call locations?

File: async_data_factory/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+                    Ok::<usize, ()>(123)
+                }
+            })
+            // a data_factory redis pool for work load comparision.
+            .data_factory(|| pool_builder(1))
+            .service(web::resource("/pool").route(web::get().to(pool_shared_prebuilt)))
+            .service(web::resource("/pool2").route(web::get().to(pool_local)))
+    })
+    .bind("127.0.0.1:8080")?
+    .workers(num_cpus)
```
Comment (robjtede) [inline async_data_factory/src/main.rs:43]: actix-web already does this if not specified

File: async_data_factory/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,92 @@
+use actix_web::web::Data;
+use actix_web::{web, App, HttpServer};
+
+use redis_tang::{Builder, Pool, RedisManager};
+
+// change according to your redis setting.
+const REDIS_URL: &str = "redis://127.0.0.1";
```
Comment (robjtede) [inline async_data_factory/src/main.rs:7]: this would be better configurable as an env var with this as default

**Comment by fakeshadow** [comment]
Changes have been made according to reviews.

File: async_data_factory/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+version = "0.1.0"
+authors = ["fakeshadow <24548779@qq.com>"]
+edition = "2018"
+workspace = ".."
+
+
+[dependencies]
+actix-rt = "1.1.1"
+actix-web = { version = "2.0.0" }
+num_cpus = "1.13.0"
```
Comment (robjtede) [inline async_data_factory/Cargo.toml:12]: nit: no longer necessary

Comment (robjtede) [reply async_data_factory/Cargo.toml:12]: also simplify actix-web line above

**Pull request merged**